core/src/main/java/com/linecorp/armeria/client/DnsResolverGroupBuilder.java

```diff
@@ -0,0 +1,345 @@
+ * meaning DNS queries after TTL will retrieve a refreshed result right away. If refreshing fails,
+ * the {@link AddressResolver} will retry with {@link #refreshBackoff(Backoff)}.
+ *
+ * <p>The refreshing {@link AddressResolver} will only start auto refresh for a given hostname
+ * on the second access before TTL to avoid auto-refreshing for queries that only happen once
+ * (e.g., requests during server startup).
+ */
+public final class DnsResolverGroupBuilder {
+
+    private Backoff refreshBackoff = Backoff.ofDefault();
+
+    private int minTtl = 1;
+    private int maxTtl = Integer.MAX_VALUE;
+    private int negativeTtl;
+
+    // DnsNameResolverBuilder properties:
+
+    private boolean traceEnabled = true;
+    @Nullable
+    private Long queryTimeoutMillis;
+    @Nullable
+    private ResolvedAddressTypes resolvedAddressTypes;
+    @Nullable
+    private Boolean recursionDesired;
+    @Nullable
+    private Integer maxQueriesPerResolve;
+    @Nullable
+    private Integer maxPayloadSize;
+    @Nullable
+    private Boolean optResourceEnabled;
+    @Nullable
+    private HostsFileEntriesResolver hostsFileEntriesResolver;
+    @Nullable
+    private DnsServerAddressStreamProvider dnsServerAddressStreamProvider;
+    @Nullable
+    private DnsQueryLifecycleObserverFactory dnsQueryLifecycleObserverFactory;
+    @Nullable
+    private List<String> searchDomains;
+    @Nullable
+    private Integer ndots;
+    @Nullable
+    private Boolean decodeIdn;
+
+    DnsResolverGroupBuilder() {}
+
+    /**
+     * Sets {@link Backoff} which is used when the {@link DnsNameResolver} fails to update the cache.
+     */
+    public DnsResolverGroupBuilder refreshBackoff(Backoff refreshBackoff) {
+        this.refreshBackoff = requireNonNull(refreshBackoff, "refreshBackoff");
+        return this;
+    }
+
+    /**
+     * Sets the minimum and maximum TTL of the cached DNS resource records in seconds. If the TTL of the DNS
+     * resource record returned by the DNS server is less than the minimum TTL or greater than the maximum TTL,
+     * this resolver will ignore the TTL from the DNS server and use the minimum TTL or the maximum TTL instead
+     * respectively.
+     * The default value is {@code 1} and {@link Integer#MAX_VALUE}, which practically tells this resolver to
+     * respect the TTL from the DNS server.
+     */
+    public DnsResolverGroupBuilder ttl(int minTtl, int maxTtl) {
+        checkArgument(minTtl > 0 && minTtl <= maxTtl,
+                      "minTtl: %s, maxTtl: %s (expected: 1 <= minTtl <= maxTtl)", minTtl, maxTtl);
+        this.minTtl = minTtl;
+        this.maxTtl = maxTtl;
+        return this;
+    }
+
+    /**
+     * Sets the TTL of the cache for the failed DNS queries in seconds. The default value is {@code 0} which
+     * means that the {@link AddressResolver} does not cache when DNS queries are failed.
+     */
+    public DnsResolverGroupBuilder negativeTtl(int negativeTtl) {
```

Comment:
`failedTtl`? Not sure what negative means here (`checkArgument(negativeTtl >= 0)` looks a little funny ;))

Also what gets cached if the query failed?

Reply:
I brought this from [netty](https://github.com/netty/netty/blob/4.1/resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java#L161)
but yes, it took a little time for me to understand what it means. We don't have to follow it. 😄 

> Also what gets cached if the query failed?

It gets the exception when the first query fails. Mostly `UnknownHostException`.

Reply:
Better using the same property name with Netty?

Reply:
I think `negative` is pretty standard for the failed cache.:-)
* https://tools.ietf.org/html/rfc2308
* https://en.wikipedia.org/wiki/Negative_cache

Reply:
Ah didn't know it's a standard term :)